### ISSUE TITLE
Add support for room peeking

### ISSUE THREAD

**Issue by akirk** [Automattic/chatrix#141]
It'd be interesting for people not having to join the chat room to see the ongoing conversation. Since Matrix has support for peeking into a room, let's research how we can implement this, ideally completely anonymous, so that only when people want to write a message, they'd log in.

**Comment by psrpinto** [comment]
Relevant hydrogen issue: https://github.com/vector-im/hydrogen-web/issues/719
PR showing how guest login could be implemented in hydrogen: https://github.com/vector-im/hydrogen-web/pull/390

**Comment by akirk** [comment]
We discussed that guest accounts are the way to go. We are unsure about the cost for them in EMS.

**Comment by ashfame** [comment]
**Update so far:**

Contribution to Hydrogen - https://github.com/ashfame/hydrogen-web/compare/master...ashfame:hydrogen-web:guest_account

**In Chatrix:** [Draft PR](https://github.com/Automattic/chatrix/pull/147)

I have added an unknown room view model and registers a guest account but at that point it fails to join the room.

![Screenshot 2022-12-14 at 12 58 59 PM](https://user-images.githubusercontent.com/858906/207551612-20c290c1-8eb2-4f41-861a-2f4ac0c8ecb5.png)

**Comment by ashfame** [comment]
**Quick update:**

I got the guest accounts functionality as I originally thought functional, but it was after joining the room with the guest account. Since, having lots of join notifications are undesirable, I was researching and was glad to learn that its certainly possible to preview room without joining the room. It just requires authentication and doesn't need to join the room. So that's what I have been working on. I am close, but I am stuck on rendering the timeline. View should re-render but it doesn't seem to.

I also researched about utilising a single guest/limited user to preview users and that's a viable route as well, though it requires some additional steps. Its possible to set power level of this limited user as -1 in every room and thus it won't be able to post a message. But it would need to be in every room that we need the preview functionality for and have a negative power level (which is the additional step required with every room). Apart from that, I don't really see any other con with this approach. The only thing malicious a user can attempt to do with this, is to join other rooms on other servers and spam there. Those servers can choose to ban that user, which is the same concern with guest account as well. Unless we block such Client Server calls using Corporal or something. Additionally, we might have to lift rate limiting for this user in that case though.

That being said, I would say, getting it first functional with guest account is the way to go. This way we have all the flexibility (someone can choose a guest account for room previews, someone can have the guest account ready for participation, and someone can designate a specific account for room previews).

**Comment by ashfame** [comment]
**Quick update:**

I am currently working on bringing the peeking functionality in Hydrogen first, so that it works for all users i.e. regular and guest accounts. This prevents it from getting too complicated to work it all within Chatrix (its codebase, changes in node modules directly for hydrogen and rebuilding vite build files). I also lost some code modifications in node_modules when I picked it back after new year vacations. There are also multiple changes in hydrogen branches, which were merged in a single branch to act as the modified dependency.

So working in hydrogen directly and landing it there is indeed a great way to move forward (thanks @psrpinto) and fact that upgrading hydrogen to the latest would simply bring the feature in, along with guest account functionality makes it a very viable route.

**Current work:**

- Guest account work in Hydrogen https://github.com/vector-im/hydrogen-web/compare/master...ashfame:hydrogen-web:guest_account?expand=1
- Guest account work in Chatrix https://github.com/Automattic/chatrix/pull/147
- Peeking work in Hydrogen https://github.com/vector-im/hydrogen-web/pull/981

**Comment by ashfame** [comment]
**Quick update:** @psrpinto and I were able to make progress on timeline by stuffing data directly in indexed db and that seems to change the view a little. I am going to continue to put more stuff in there and see if the timeline gets rendered that way.

We also evaluated matrix-public-archive and that works by having the user (whose access token it uses) to be joined to the room.

We also discussed options of bringing peeking functionality since we both find it quite valuable and decided that continuing on current path of landing peeking support in hydrogen is indeed the best path forward.

My hope is that, with more stuff put into indexed db, we would atleast see the timeline rendered properly and it would just be the sync that we would need to write some code for, to keep showing new messages as they come into the room.

**Comment by akirk** [comment]
Great, thanks for the update!

**Comment by ashfame** [comment]
**Quick update:** Finally got the timeline to render during today's pairing session with @psrpinto 
Still have rough edges to fix but definitely a milestone 🥳 

Things that are yet to do:
1) Reliably sync new messages into the indexed db store. Second message in the room showed up when refreshed but third one didn't
2) Messages show up in the wrong order
3) Room name doesn't show up yet
4) Styling of bottom bar
5) Actual sync mechanism (which is to be tackled in a later PR)

**Comment by ashfame** [comment]
Draft PR link for review - https://github.com/Automattic/hydrogen-web/pull/2

**Comment by akirk** [comment]
> Actual sync mechanism (which is to be tackled in a later PR)

Yes, I think this is not urgent and I am not sure it is actually implemented even in Element itself.

**Comment by psrpinto** [comment]
I agree that having new messages come in after the timeline is loaded is something we could look at in the future, once/if "static" peaking gets merged into hydrogen.

**Comment by ashfame** [comment]
Yep, agree sync can come later on. For sake of documentation, Element does support it.

I have made the following changes:

- Syncing happens correctly irrespective of previous loaded state when page is reloaded
- All messages now show up correctly
- Changed the order of fetching messages to be chronological
- Room name shows up now
- Context menu on messages are now hidden

What still needs doing:

- Load members info (currently it shows user identity in absence of that)
- Back scrolling doesn't work (its only the initial messages that are loaded)
- Bottom bar styling

**Comment by ashfame** [comment]
Got a bunch of more things done today:

- Refactor the function to fetch events so that it can be easily invoked multiple times (for future needs)
- Generate room summary dynamically
- move peek() invocation outside of constructor
- Style bottom bar
- Room's picture now shows up
- Load 100 messages to begin with, since not focusing on making back scrolling functional

Couple of things that I would say are remaining:

- Load members info (user name and picture)
- Disable right panel view rendering when clicked on user picture

**Comment by ashfame** [comment]
Draft PR for Hydrogen - https://github.com/Automattic/hydrogen-web/pull/3

Members identity are shown instead of names. Apparently loading members list brings in more undesirable effects so that has been left for later. Peeking is functional! 🥳 

Getting it reviewed internally before submitting it to upstream.

**Comment by ashfame** [comment]
Updated original comment with upstream PR link and next step to follow.

**Comment by ashfame** [comment]
### Grand update:
In order to have peeking support in Chatrix, there are 2 user stories to satisfy.

#### User Story 1

I am already a logged in user on Matrix homeserver and I am browsing different pages where different rooms are embedded. On one of those page, I can land on a room that I haven't joined. But if that room is world readable, I can just peek into it, using my existing account.

This user story is satisfied by the [upstream PR for peeking](https://github.com/vector-im/hydrogen-web/pull/1037), which is about just using an existing session to peek into an unknown room if its world readable.

#### User Story 2

I am not logged in on Matrix homeserver at all and I am browsing pages where a matrix room is embedded. If any of the rooms are world readable, I should still be able to peek into the room, by having a guest account created for myself behind the scenes.

This user story would be satisfied once upstream PR for peeking is merged and another PR for creating a guest account when trying to render Chatrix in single room mode which would work as following:

- Introduce the possibility of rendering an unknown room, even without a session instead of redirecting to login screen (specific to Chatrix)
- While trying to render this unknown room, we will check whether room is world_readable and if so, create a guest session if there isn't already one (`client.doGuestLogin()` would be contributed to Hydrogen, rest would be specific to Chatrix)
- Finally just redirect the user to `/session/XXXXX/room/YYYY` view and it would just work as peeking into a peekable room with a user account's session you have.

##### Existing work for User story 2

Work in following 2 links would need to be adapted on top of our upstream peeking PR:

- https://github.com/Automattic/chatrix/pull/147/files
- https://github.com/vector-im/hydrogen-web/compare/master...ashfame:hydrogen-web:guest_account?expand=1

I tried running peeking PR branch as Hydrogen dependency in Chatrix, to prepare how this second PR would look like but there are some issues that we first need to resolve prior to update our dependency to Hydrogen version `v0.3.7` (we use `v0.3.5` and latest is `v0.3.8`)

#### PS

Once we attempt to package "Guest as login" proposal to Hydrogen as whole, some of the specific code for user story 2, sitting just in Chatrix could potentially be contributed upstream. Not all of it would be possible, since "Single Room Mode" is a Chatrix's concept.

**Comment by ashfame** [comment]
First comment of issue updated with links to relevant PRs to finish this feature.
